src/drivers/oreoled/oreoled_bootloader/oreoled_bootloader_avr.cpp

```diff
@@ -0,0 +1,1202 @@
+	memset(_healthy, 0, sizeof(_healthy));
+
+	/* initialise to in application */
+	memset(_in_boot, 0, sizeof(_in_boot));
+
+	/* capture startup time */
+	_start_time = hrt_absolute_time();
+}
+
+/* destructor */
+OREOLED_BOOTLOADER_AVR::~OREOLED_BOOTLOADER_AVR()
+{
+	/* free perf counters */
+	perf_free(_probe_perf);
+	perf_free(_comms_errors);
+	perf_free(_reply_errors);
+}
+
+int
+OREOLED_BOOTLOADER_AVR::start(void)
+{
+	int ret;
+
+	/* initialise I2C bus */
+	ret = I2C::init();
+
+	if (ret != OK) {
+		return ret;
+	}
+
+	startup_discovery();
+
+	return OK;
+}
+
+int
+OREOLED_BOOTLOADER_AVR::update(bool force)
+{
+	int ret;
+
+	ret = start();
+	if (ret != OK) {
+		return ret;
+	}
+
+	run_updates(force);
+
+	print_info();
+
+	return OK;
+}
+
+int
+OREOLED_BOOTLOADER_AVR::ioctl(const unsigned cmd, const unsigned long arg)
+{
+	int ret = EINVAL;
+
+	switch (cmd) {
+	case OREOLED_BL_RESET:
+		ret = app_reset_all();
+		return ret;
+
+	case OREOLED_BL_PING:
+		for (uint8_t i = 0; i < OREOLED_NUM_LEDS; i++) {
+			if (_healthy[i]) {
+				ping(i);
+				ret = OK;
+			}
+		}
+		return ret;
+
+	case OREOLED_BL_VER:
+		for (uint8_t i = 0; i < OREOLED_NUM_LEDS; i++) {
+			if (_healthy[i]) {
+				version(i);
+				ret = OK;
+			}
+		}
+		return ret;
+
+	case OREOLED_BL_FLASH:
+		ret = flash_all(true);
+		return ret;
+
+	case OREOLED_BL_APP_VER:
+		for (uint8_t i = 0; i < OREOLED_NUM_LEDS; i++) {
+			if (_healthy[i]) {
+				app_version(i);
+				ret = OK;
+			}
+		}
+		return ret;
+
+	case OREOLED_BL_APP_CHECKSUM:
+		for (uint8_t i = 0; i < OREOLED_NUM_LEDS; i++) {
+			if (_healthy[i]) {
+				app_checksum(i);
+				ret = OK;
+			}
+		}
+		return ret;
+
+	case OREOLED_BL_SET_COLOUR:
+		for (uint8_t i = 0; i < OREOLED_NUM_LEDS; i++) {
+			if (_healthy[i]) {
+				set_colour(i, ((oreoled_rgbset_t *) arg)->red, ((oreoled_rgbset_t *) arg)->green);
+				ret = OK;
+			}
+		}
+		return ret;
+
+	case OREOLED_BL_BOOT_APP:
+		ret = boot_all();
+		return ret;
+	}
+
+	return ret;
+}
+
+void
+OREOLED_BOOTLOADER_AVR::print_info(void)
+{
+	/* print health info on each LED */
+	for (uint8_t i = 0; i < OREOLED_NUM_LEDS; i++) {
+		if (!_healthy[i]) {
+			warnx("oreo %u: BAD", (unsigned)i);
+
+		} else {
+			warnx("oreo %u: OK", (unsigned)i);
+		}
+	}
+
+	/* display perf info */
+	perf_print_counter(_probe_perf);
+	perf_print_counter(_comms_errors);
+	perf_print_counter(_reply_errors);
+}
+
+void
+OREOLED_BOOTLOADER_AVR::startup_discovery(void)
+{
+	/* check time since startup */
+	uint64_t now = hrt_absolute_time();
+	bool startup_timeout = (now - _start_time > OREOLED_STARTUP_TIMEOUT_USEC);
+
+	/* during startup period keep searching for unhealthy LEDs */
+	do {
+		discover();
+
+		if (_num_healthy == OREOLED_NUM_LEDS) {
+			break;
+		}
+
+		/* wait for 0.1 sec before running discovery again */
+		usleep(OREOLED_STARTUP_INTERVAL_US);
+	} while ((hrt_absolute_time() - _start_time) < OREOLED_STARTUP_TIMEOUT_USEC);
+}
+
+void
+OREOLED_BOOTLOADER_AVR::discover(void)
+{
```

Comment:
How do you envision discovery working once new bootloader types are added? This probe request probably belongs in the core driver before the driver backend is chosen.